lib/devicetrust/authz/authz_test.go

```diff
@@ -238,3 +258,60 @@ func runVerifyUserTest(t *testing.T, method string, verify func(dt *types.Device
 		})
 	}
 }
+
+func TestCalculateTrustedDeviceRequirement(t *testing.T) {
+	deviceTrustOptionalRole, err := types.NewRole("editor", types.RoleSpecV6{Options: types.RoleOptions{DeviceTrustMode: constants.DeviceTrustModeOptional}})
+	assert.NoError(t, err)
+	deviceTrustRequiredRole, err := types.NewRole("editor", types.RoleSpecV6{Options: types.RoleOptions{DeviceTrustMode: constants.DeviceTrustModeRequired}})
+	assert.NoError(t, err)
+
+	tests := []struct {
+		name              string
+		dt                *types.DeviceTrust
+		roles             []types.Role
+		expectRequirement types.TrustedDeviceRequirement
+	}{
+		{
+			name: "not required by cluster or by roles",
+			dt: &types.DeviceTrust{
+				Mode: constants.DeviceTrustModeOff,
+			},
+			roles:             []types.Role{deviceTrustOptionalRole},
+			expectRequirement: types.TrustedDeviceRequirement_TRUSTED_DEVICE_REQUIREMENT_NOT_REQUIRED,
+		},
+		{
+			name: "not required by cluster or by roles",
+			dt: &types.DeviceTrust{
+				Mode: "",
+			},
+			roles:             []types.Role{deviceTrustOptionalRole},
+			expectRequirement: types.TrustedDeviceRequirement_TRUSTED_DEVICE_REQUIREMENT_NOT_REQUIRED,
+		},
+		{
+			name: "required by cluster but not by roles",
+			dt: &types.DeviceTrust{
+				Mode: constants.DeviceTrustModeRequired,
+			},
+			roles:             []types.Role{deviceTrustOptionalRole},
+			expectRequirement: types.TrustedDeviceRequirement_TRUSTED_DEVICE_REQUIREMENT_REQUIRED,
+		},
+		{
+			name: "required by role but not by cluster",
+			dt: &types.DeviceTrust{
+				Mode: constants.DeviceTrustModeRequired,
+			},
+			roles:             []types.Role{deviceTrustRequiredRole},
```

Comment:
Suggestion: Might as well check with both roles here:

```suggestion
			roles:             []types.Role{deviceTrustOptionalRole, deviceTrustRequiredRole},
```

We are lacking a test with >1 role in the set anyway.